src/Processors/Transforms/DeduplicationTokenTransforms.cpp

```diff
@@ -0,0 +1,241 @@
+
+    stage = DEFINED;
+}
+
+void TokenInfo::setUserToken(const String & token)
+{
+    if (stage == UNDEFINED)
+        stage = DEFINE_SOURCE_USER_TOKEN;
+
+    if (stage != DEFINE_SOURCE_USER_TOKEN)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    addTokenPart(fmt::format("user-token-{}", token));
+}
+
+void TokenInfo::defineSourceWithUserToken(size_t block_number)
+{
+    if (stage != DEFINE_SOURCE_USER_TOKEN)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    addTokenPart(fmt::format("source-number-{}", block_number));
+
+    stage = DEFINED;
+}
+
+void TokenInfo::setViewID(const String & id)
+{
+    if (stage == DEFINED)
+        stage = DEFINE_VIEW;
+
+    if (stage != DEFINE_VIEW)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    addTokenPart(fmt::format("view-id-{}", id));
+}
+
+void TokenInfo::defineViewID(size_t block_number)
+{
+    if (stage != DEFINE_VIEW)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is in wrong stage {}, token {}", stage, debugToken());
+
+    addTokenPart(fmt::format("view-block-{}", block_number));
+
+    stage = DEFINED;
+}
+
+void TokenInfo::reset()
+{
+    stage = UNDEFINED;
+    parts.clear();
+}
+
+void TokenInfo::addTokenPart(String part)
+{
+    parts.push_back(std::move(part));
+}
+
+size_t TokenInfo::getTotalSize() const
+{
+    if (parts.empty())
+        return 0;
+
+    size_t size = 0;
+    for (const auto & part : parts)
+        size += part.size();
+
+    return size + parts.size() - 1;
+}
+
+#ifdef ABORT_ON_LOGICAL_ERROR
+void CheckTokenTransform::transform(Chunk & chunk)
+{
+    auto token_info = chunk.getChunkInfos().get<TokenInfo>();
+
+    if (!token_info)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Chunk has to have DedupTokenInfo as ChunkInfo, {}", debug);
+
+    if (!must_be_present)
+    {
+        LOG_DEBUG(log, "{}, no token required, token {}", debug, token_info->debugToken());
+        return;
+    }
+
+    LOG_DEBUG(log, "debug: {}, token: {}", debug, token_info->debugToken());
+}
+#endif
+
+String SetInitialTokenTransform::getChunkHash(const Chunk & chunk)
+{
+    SipHash hash;
+    for (const auto & colunm : chunk.getColumns())
+        colunm->updateHashFast(hash);
+
+    const auto hash_value = hash.get128();
+    return toString(hash_value.items[0]) + "_" + toString(hash_value.items[1]);
+}
+
+
+void SetInitialTokenTransform::transform(Chunk & chunk)
+{
+    auto token_info = chunk.getChunkInfos().get<TokenInfo>();
+
+    if (!token_info)
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "TokenInfo is expected for consumed chunk in SetInitialTokenTransform");
+
+    if (token_info->isDefined())
+        return;
+
+    token_info->addChunkHash(getChunkHash(chunk));
+    token_info->defineSourceWithChunkHashes();
```

Comment:
I got an impression initially that we can call `getChunkHash` several times and then do `defineSourceWithChunkHashes`, but here you do the second staight after the first, so stage is `DEFINED` after adding the first chunk hash. Am I missing something?

Reply:
add a comment to this cals.
renamed tokens methods.
Here is only one hash calculated. It is a hash from entire chunk. Usually [Replicated]MergeTreeSink adds a hashed to the deduplication token as they write parts to the disk. But when table has a different engine we still have to define the source in token.